Source/Core/AsyncQueue.cs

```diff
@@ -0,0 +1,70 @@
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Boogie;
+
+/// <summary>
+/// A queue from which items can be dequeued even if they have not been enqueued yet.
+/// All methods are thread-safe.
+/// Also supports pushing elements to the front of the queue.
+/// </summary>
+public class AsyncQueue<T>
+{
+  private readonly LinkedList<T> items = new();
+  private readonly Queue<TaskCompletionSource<T>> customers = new();
+
+  public void Enqueue(T value)
+  {
+    lock (this) {
+      while (customers.TryDequeue(out var customer)) {
+        if (customer.TrySetResult(value)) {
```

Comment:
We might to have to be careful with `TaskCompletionSource`
See here for example.
https://devblogs.microsoft.com/premier-developer/the-danger-of-taskcompletionsourcet-class/

In short, it will invoke the continuation attached to customers.... synchronously. Hence, what happens after `await asyncQueue.Dequeue()` until the next `await` will executed during this call, which might take some time. I looked statically at the hypothetical trace, it would go back to `DoWork`, then `StartCheck`, then `BeginCheck`, then a lot of computation to prepare the next VC condition... and during all that time, the queue is locked and released checkers cannot be used by other tasks.

The solution here is to lock only extracting a customer.

Reply:
Replied in your other comment.